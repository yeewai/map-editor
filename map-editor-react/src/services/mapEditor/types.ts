import { structureDefinitionTypes } from 'services/structureDefinitions';
import { worldTypes } from 'services/worlds';

export interface State {
    zoom: number,
    panV: number,
    panH: number,
    buildStructureDefinition?: structureDefinitionTypes.StructureDefinition
    mousePosition?: { x: number, y: number },
    activeWorld?: worldTypes.World
}
