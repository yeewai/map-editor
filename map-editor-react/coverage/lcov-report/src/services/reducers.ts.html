<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/services/reducers.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">src/services</a> reducers.ts
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>0/10</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/4</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/1</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>0/5</span>
      </div>
    </div>
  </div>
  <div class='status-line low'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13</td><td class="line-coverage quiet"><span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import { rout<span class="cstat-no" title="statement not covered" >erReducer } from 'react-route<span class="cstat-no" title="statement not covered" >r</span>-redux';<span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ></span></span></span>
import { comb<span class="cstat-no" title="statement not covered" >ineReducers } fro</span>m 'redux';
<span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="fstat-no" title="function not covered" ></span></span></span></span>
import { StateTree<span class="cstat-no" title="statement not covered" > } from './types';</span>
import structureDefinitions from './structureDefinitions';
<span class="cstat-no" title="statement not covered" ></span>
const rootReducer = combineReducers&lt;StateTree&gt;({
    router: routerReducer,
    structureDefinitions
});
&nbsp;
export default rootReducer;
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Jan 31 2018 09:59:49 GMT-0600 (CST)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
