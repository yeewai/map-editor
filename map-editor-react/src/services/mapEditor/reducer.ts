import { Action } from 'services/types';
import * as Types from './types';
import { worldTypes } from 'services/worlds';

export const defaultState: Types.State = {
    zoom: 1,
    panV: 0,
    panH: 0,
    buildStructureDefinition: undefined,
    mousePosition: undefined,
    activeWorld: undefined
}

export const ZOOMRATE = 1.5;
export const MAXZOOM = 10;
export const PANRATE = 1;

const mapEditorReducer = (state = defaultState, action: Action ): Types.State => {
    switch(action.type) {
        case 'mapEditor/SET_ACTIVE':
            return { ...state, activeWorld: action.payload };
        case 'mapEditor/ZOOMIN':
            return { ...state, zoom: Math.min(state.zoom * ZOOMRATE, MAXZOOM) };
        case 'mapEditor/ZOOMOUT':
            return { ...state, zoom: state.zoom / ZOOMRATE };
        case 'mapEditor/ZOOMRESET':
            return { ...state, zoom: defaultState.zoom };
        case 'mapEditor/PANUP':
            return {...state, panV: state.panV - PANRATE}
        case 'mapEditor/PANDOWN':
            return {...state, panV: state.panV + PANRATE}
        case 'mapEditor/PANLEFT':
            return {...state, panH: state.panH - PANRATE}
        case 'mapEditor/PANRIGHT':
            return {...state, panH: state.panH + PANRATE}
        case 'mapEditor/PANRESET':
            return {...state, panH: defaultState.panH, panV: defaultState.panV}
        case 'mapEditor/SET_BUILD_STRUCTUREDEFINITION':
            const sd = action.payload === state.buildStructureDefinition
                ? defaultState.buildStructureDefinition
                : action.payload;
            return {...state, buildStructureDefinition: sd}
        // Will probably use this for keyboard controls
        // case 'mapEditor/UNSET_BUILD_STRUCTUREDEFINITION':
        //     return {...state, buildStructureDefinition: defaultState.buildStructureDefinition}
        //
        case 'mapEditor/SET_MOUSE_POSITION':
            return {...state, mousePosition: action.payload }

        case 'mapEditor/MOUSE_CLICK':
            const { mousePosition, buildStructureDefinition } = state;
            if (!(mousePosition && buildStructureDefinition)) { return state; }

            const newStructure: worldTypes.Structure = {
                ...worldTypes.emptyStructure,
                name: "",
                definitionId: buildStructureDefinition.id,
                definition: buildStructureDefinition,
                xposition: mousePosition.x ,
                yposition: mousePosition.y
            }

            return {
                ...state,
                activeWorld: {
                    ...state.activeWorld,
                    structures: [
                        ...state.activeWorld.structures || [],
                        newStructure
                    ]
                }
            };

        default:
            return state;
    }
}


export default mapEditorReducer;
