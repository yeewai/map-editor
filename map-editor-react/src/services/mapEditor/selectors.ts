import { createSelector } from 'reselect';
import _ from 'lodash';
import { StateTree } from 'services/types';
import { structureDefinitionsSelectors, structureDefinitionTypes as sdTypes } from 'services/structureDefinitions';
import { worldTypes } from 'services/worlds';

// -------------------------------------------------------------
// Camera Controls
// -------------------------------------------------------------
export const getZoom = ( state: StateTree ) => state.mapEditor.zoom;
export const getPan = (state: StateTree ) => ({ h: state.mapEditor.panH, v: state.mapEditor.panV });

export const getBuildStructureDefinition = ( state:StateTree ) => state.mapEditor.buildStructureDefinition;
export const getMousePosition = ( state:StateTree ) => state.mapEditor.mousePosition;

// -------------------------------------------------------------
// Map Editor Active World
// -------------------------------------------------------------
export const getActiveWorld = ( state: StateTree ) => state.mapEditor.activeWorld;

export const getActiveWorldWithDefinition = createSelector(
    getActiveWorld, structureDefinitionsSelectors.get,
    ( world: worldTypes.World, sdefinitions: sdTypes.StructureDefinition[] ) => {
        if (!world) { return null; }
        const structures = world.structures && world.structures.map( s => ({
            ...s,
            withDefinition: true,
            definition: sdefinitions.find( sd => sd.id === s.definitionId)
        }))
        return {
            ...world,
            structures
        };
    }
)

export const getBoardFromWorld = ( world: worldTypes.World ): worldTypes.Structure[][] => {
    //Mutable, but I think it's more efficient tbh...
    let board: worldTypes.Structure[][] = _.range(world.width).map( (x, i) => (_.range(world.length).map( (y, j) => (null))));

    if (world.structures) {
        world.structures.forEach( s => {
            board[s.xposition][s.yposition] = s
        })
    }

    return board;
}

export const getUniqueStructuresInWorld = ( world: worldTypes.World): sdTypes.StructureDefinition[] => {
    return (
        world.structures
        ? Array.from(new Set(world.structures.map( s => s.definition )))
        : []
    );
}

export const getActiveWorldWithBoard = createSelector(
    getActiveWorldWithDefinition, structureDefinitionsSelectors.get,
    ( world: worldTypes.World, sdefinitions: sdTypes.StructureDefinition[] ): worldTypes.WorldWithBoard => {
        if (!world) { return null; }

        return {
            ...world,
            board: getBoardFromWorld(world),
            uniqueStructures: getUniqueStructuresInWorld(world)
        }
    }
)
