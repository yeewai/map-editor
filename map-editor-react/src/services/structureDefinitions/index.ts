import reducer from "./reducer";
import * as structureDefinitionsActions from "./actions";
import * as structureDefinitionsSelectors from "./selectors";
import * as structureDefinitionTypes from "./types";

export { structureDefinitionsActions, structureDefinitionsSelectors, structureDefinitionTypes };
export default reducer;
