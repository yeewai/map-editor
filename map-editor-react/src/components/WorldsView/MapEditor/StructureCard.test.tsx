import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';

import { getMockRouterProps } from 'helpers/test/react-router';
import { StructureCard, mapStateToProps, mapDispatchToProps } from './StructureCard';
// import { worldTypes } from 'services/worlds';
import { defaultStateTree } from 'services/types';
import { mapEditorSelectors, mapEditorActions } from 'services/mapEditor';

Enzyme.configure({ adapter: new Adapter() });

describe("Map Editor", () => {
    let sandbox, wrapper, routerProps, world;

    beforeEach(() => {
        sandbox = sinon.createSandbox();
        routerProps = getMockRouterProps<any>(null);
        // world = worldTypes.emptyWorldWithBoard;
    });

    afterEach(() => {
        sandbox.restore();
    });

    it ("renders", () => {
        wrapper = shallow(<StructureCard structureDefinition={{name: "a", width: 1, length: 2, description: "<p>Sup</p>"}} isActive={true} setBuildStructureDefinition={sandbox.spy()} />);
        expect(wrapper).toMatchSnapshot();
    });

    it ("renders an inactive structure definition", () => {
        wrapper = shallow(<StructureCard structureDefinition={{name: "a", width: 1, length: 2}} isActive={false} setBuildStructureDefinition={sandbox.spy()} />);
        expect(wrapper).toMatchSnapshot();
    });

    it ("calls setBuildStructureDefinition on click", () => {
        const spy = sandbox.spy();
        const wrapper = shallow(<StructureCard structureDefinition={} isActive={false} setBuildStructureDefinition={spy()} />);
        wrapper.simulate("click");

        expect(spy.called).toBeTruthy();
    })


    it ("maps state to props", () => {
        const ownProps = { structureDefinition: "a"};
        sandbox.stub(mapEditorSelectors, 'getBuildStructureDefinition').returns("a")

        expect(mapStateToProps(defaultStateTree, ownProps )).toBeTruthy()
    })

    it ("handles setting the build structure", () => {
        const spy = sandbox.stub(mapEditorActions, 'setStructureDefinition');
        const dispatchSpy = sandbox.spy();

        mapDispatchToProps(dispatchSpy).setBuildStructureDefinition("a");
        expect(spy.called).toBe(true);

        sandbox.restore();
    })

});
