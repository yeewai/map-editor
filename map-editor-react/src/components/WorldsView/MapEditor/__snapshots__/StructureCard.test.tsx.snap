// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map Editor renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    isActive={true}
    setBuildStructureDefinition={[Function]}
    structureDefinition={
        Object {
            "description": "<p>Sup</p>",
            "length": 2,
            "name": "a",
            "width": 1,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        " ",
        <button
          className="structure-btn"
          onClick={[Function]}
>
          <Card
                    tag="div"
          >
                    <CardImg
                              alt="a"
                              src={undefined}
                              tag="img"
                              top={true}
                              width="100%"
                    />
                    <CardBody
                              tag="div"
                    >
                              <CardTitle
                                        tag="h4"
                              >
                                        a
                              </CardTitle>
                              <CardSubtitle
                                        tag="h6"
                              >
                                        1
                                         x 
                                        2
                              </CardSubtitle>
                              <div
                                        className="card-text"
                                        dangerouslySetInnerHTML={
                                                  Object {
                                                            "__html": "<p>Sup</p>",
                                                          }
                                        }
                              />
                    </CardBody>
          </Card>
</button>,
      ],
      "className": "card-li active",
    },
    "ref": null,
    "rendered": Array [
      " ",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Card
            tag="div"
>
            <CardImg
                        alt="a"
                        src={undefined}
                        tag="img"
                        top={true}
                        width="100%"
            />
            <CardBody
                        tag="div"
            >
                        <CardTitle
                                    tag="h4"
                        >
                                    a
                        </CardTitle>
                        <CardSubtitle
                                    tag="h6"
                        >
                                    1
                                     x 
                                    2
                        </CardSubtitle>
                        <div
                                    className="card-text"
                                    dangerouslySetInnerHTML={
                                                Object {
                                                            "__html": "<p>Sup</p>",
                                                          }
                                    }
                        />
            </CardBody>
</Card>,
          "className": "structure-btn",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <CardImg
                alt="a"
                src={undefined}
                tag="img"
                top={true}
                width="100%"
/>,
              <CardBody
                tag="div"
>
                <CardTitle
                                tag="h4"
                >
                                a
                </CardTitle>
                <CardSubtitle
                                tag="h6"
                >
                                1
                                 x 
                                2
                </CardSubtitle>
                <div
                                className="card-text"
                                dangerouslySetInnerHTML={
                                                Object {
                                                                "__html": "<p>Sup</p>",
                                                              }
                                }
                />
</CardBody>,
            ],
            "tag": "div",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "alt": "a",
                "src": undefined,
                "tag": "img",
                "top": true,
                "width": "100%",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <CardTitle
                    tag="h4"
>
                    a
</CardTitle>,
                  <CardSubtitle
                    tag="h6"
>
                    1
                     x 
                    2
</CardSubtitle>,
                  <div
                    className="card-text"
                    dangerouslySetInnerHTML={
                                        Object {
                                                            "__html": "<p>Sup</p>",
                                                          }
                    }
/>,
                ],
                "tag": "div",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": "a",
                    "tag": "h4",
                  },
                  "ref": null,
                  "rendered": "a",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": Array [
                      1,
                      " x ",
                      2,
                    ],
                    "tag": "h6",
                  },
                  "ref": null,
                  "rendered": Array [
                    1,
                    " x ",
                    2,
                  ],
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "card-text",
                    "dangerouslySetInnerHTML": Object {
                      "__html": "<p>Sup</p>",
                    },
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "div",
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": "button",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          " ",
          <button
            className="structure-btn"
            onClick={[Function]}
>
            <Card
                        tag="div"
            >
                        <CardImg
                                    alt="a"
                                    src={undefined}
                                    tag="img"
                                    top={true}
                                    width="100%"
                        />
                        <CardBody
                                    tag="div"
                        >
                                    <CardTitle
                                                tag="h4"
                                    >
                                                a
                                    </CardTitle>
                                    <CardSubtitle
                                                tag="h6"
                                    >
                                                1
                                                 x 
                                                2
                                    </CardSubtitle>
                                    <div
                                                className="card-text"
                                                dangerouslySetInnerHTML={
                                                            Object {
                                                                        "__html": "<p>Sup</p>",
                                                                      }
                                                }
                                    />
                        </CardBody>
            </Card>
</button>,
        ],
        "className": "card-li active",
      },
      "ref": null,
      "rendered": Array [
        " ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Card
              tag="div"
>
              <CardImg
                            alt="a"
                            src={undefined}
                            tag="img"
                            top={true}
                            width="100%"
              />
              <CardBody
                            tag="div"
              >
                            <CardTitle
                                          tag="h4"
                            >
                                          a
                            </CardTitle>
                            <CardSubtitle
                                          tag="h6"
                            >
                                          1
                                           x 
                                          2
                            </CardSubtitle>
                            <div
                                          className="card-text"
                                          dangerouslySetInnerHTML={
                                                        Object {
                                                                      "__html": "<p>Sup</p>",
                                                                    }
                                          }
                            />
              </CardBody>
</Card>,
            "className": "structure-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <CardImg
                  alt="a"
                  src={undefined}
                  tag="img"
                  top={true}
                  width="100%"
/>,
                <CardBody
                  tag="div"
>
                  <CardTitle
                                    tag="h4"
                  >
                                    a
                  </CardTitle>
                  <CardSubtitle
                                    tag="h6"
                  >
                                    1
                                     x 
                                    2
                  </CardSubtitle>
                  <div
                                    className="card-text"
                                    dangerouslySetInnerHTML={
                                                      Object {
                                                                        "__html": "<p>Sup</p>",
                                                                      }
                                    }
                  />
</CardBody>,
              ],
              "tag": "div",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "alt": "a",
                  "src": undefined,
                  "tag": "img",
                  "top": true,
                  "width": "100%",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": Array [
                    <CardTitle
                      tag="h4"
>
                      a
</CardTitle>,
                    <CardSubtitle
                      tag="h6"
>
                      1
                       x 
                      2
</CardSubtitle>,
                    <div
                      className="card-text"
                      dangerouslySetInnerHTML={
                                            Object {
                                                                  "__html": "<p>Sup</p>",
                                                                }
                      }
/>,
                  ],
                  "tag": "div",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "children": "a",
                      "tag": "h4",
                    },
                    "ref": null,
                    "rendered": "a",
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "children": Array [
                        1,
                        " x ",
                        2,
                      ],
                      "tag": "h6",
                    },
                    "ref": null,
                    "rendered": Array [
                      1,
                      " x ",
                      2,
                    ],
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "card-text",
                      "dangerouslySetInnerHTML": Object {
                        "__html": "<p>Sup</p>",
                      },
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "div",
                  },
                ],
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": "button",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Map Editor renders an inactive structure definition 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    isActive={false}
    setBuildStructureDefinition={[Function]}
    structureDefinition={
        Object {
            "length": 2,
            "name": "a",
            "width": 1,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        " ",
        <button
          className="structure-btn"
          onClick={[Function]}
>
          <Card
                    tag="div"
          >
                    <CardImg
                              alt="a"
                              src={undefined}
                              tag="img"
                              top={true}
                              width="100%"
                    />
                    <CardBody
                              tag="div"
                    >
                              <CardTitle
                                        tag="h4"
                              >
                                        a
                              </CardTitle>
                              <CardSubtitle
                                        tag="h6"
                              >
                                        1
                                         x 
                                        2
                              </CardSubtitle>
                              <div
                                        className="card-text"
                                        dangerouslySetInnerHTML={
                                                  Object {
                                                            "__html": "",
                                                          }
                                        }
                              />
                    </CardBody>
          </Card>
</button>,
      ],
      "className": "card-li null",
    },
    "ref": null,
    "rendered": Array [
      " ",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Card
            tag="div"
>
            <CardImg
                        alt="a"
                        src={undefined}
                        tag="img"
                        top={true}
                        width="100%"
            />
            <CardBody
                        tag="div"
            >
                        <CardTitle
                                    tag="h4"
                        >
                                    a
                        </CardTitle>
                        <CardSubtitle
                                    tag="h6"
                        >
                                    1
                                     x 
                                    2
                        </CardSubtitle>
                        <div
                                    className="card-text"
                                    dangerouslySetInnerHTML={
                                                Object {
                                                            "__html": "",
                                                          }
                                    }
                        />
            </CardBody>
</Card>,
          "className": "structure-btn",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <CardImg
                alt="a"
                src={undefined}
                tag="img"
                top={true}
                width="100%"
/>,
              <CardBody
                tag="div"
>
                <CardTitle
                                tag="h4"
                >
                                a
                </CardTitle>
                <CardSubtitle
                                tag="h6"
                >
                                1
                                 x 
                                2
                </CardSubtitle>
                <div
                                className="card-text"
                                dangerouslySetInnerHTML={
                                                Object {
                                                                "__html": "",
                                                              }
                                }
                />
</CardBody>,
            ],
            "tag": "div",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "alt": "a",
                "src": undefined,
                "tag": "img",
                "top": true,
                "width": "100%",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <CardTitle
                    tag="h4"
>
                    a
</CardTitle>,
                  <CardSubtitle
                    tag="h6"
>
                    1
                     x 
                    2
</CardSubtitle>,
                  <div
                    className="card-text"
                    dangerouslySetInnerHTML={
                                        Object {
                                                            "__html": "",
                                                          }
                    }
/>,
                ],
                "tag": "div",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": "a",
                    "tag": "h4",
                  },
                  "ref": null,
                  "rendered": "a",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": Array [
                      1,
                      " x ",
                      2,
                    ],
                    "tag": "h6",
                  },
                  "ref": null,
                  "rendered": Array [
                    1,
                    " x ",
                    2,
                  ],
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "card-text",
                    "dangerouslySetInnerHTML": Object {
                      "__html": "",
                    },
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "div",
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": "button",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          " ",
          <button
            className="structure-btn"
            onClick={[Function]}
>
            <Card
                        tag="div"
            >
                        <CardImg
                                    alt="a"
                                    src={undefined}
                                    tag="img"
                                    top={true}
                                    width="100%"
                        />
                        <CardBody
                                    tag="div"
                        >
                                    <CardTitle
                                                tag="h4"
                                    >
                                                a
                                    </CardTitle>
                                    <CardSubtitle
                                                tag="h6"
                                    >
                                                1
                                                 x 
                                                2
                                    </CardSubtitle>
                                    <div
                                                className="card-text"
                                                dangerouslySetInnerHTML={
                                                            Object {
                                                                        "__html": "",
                                                                      }
                                                }
                                    />
                        </CardBody>
            </Card>
</button>,
        ],
        "className": "card-li null",
      },
      "ref": null,
      "rendered": Array [
        " ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Card
              tag="div"
>
              <CardImg
                            alt="a"
                            src={undefined}
                            tag="img"
                            top={true}
                            width="100%"
              />
              <CardBody
                            tag="div"
              >
                            <CardTitle
                                          tag="h4"
                            >
                                          a
                            </CardTitle>
                            <CardSubtitle
                                          tag="h6"
                            >
                                          1
                                           x 
                                          2
                            </CardSubtitle>
                            <div
                                          className="card-text"
                                          dangerouslySetInnerHTML={
                                                        Object {
                                                                      "__html": "",
                                                                    }
                                          }
                            />
              </CardBody>
</Card>,
            "className": "structure-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <CardImg
                  alt="a"
                  src={undefined}
                  tag="img"
                  top={true}
                  width="100%"
/>,
                <CardBody
                  tag="div"
>
                  <CardTitle
                                    tag="h4"
                  >
                                    a
                  </CardTitle>
                  <CardSubtitle
                                    tag="h6"
                  >
                                    1
                                     x 
                                    2
                  </CardSubtitle>
                  <div
                                    className="card-text"
                                    dangerouslySetInnerHTML={
                                                      Object {
                                                                        "__html": "",
                                                                      }
                                    }
                  />
</CardBody>,
              ],
              "tag": "div",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "alt": "a",
                  "src": undefined,
                  "tag": "img",
                  "top": true,
                  "width": "100%",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": Array [
                    <CardTitle
                      tag="h4"
>
                      a
</CardTitle>,
                    <CardSubtitle
                      tag="h6"
>
                      1
                       x 
                      2
</CardSubtitle>,
                    <div
                      className="card-text"
                      dangerouslySetInnerHTML={
                                            Object {
                                                                  "__html": "",
                                                                }
                      }
/>,
                  ],
                  "tag": "div",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "children": "a",
                      "tag": "h4",
                    },
                    "ref": null,
                    "rendered": "a",
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "children": Array [
                        1,
                        " x ",
                        2,
                      ],
                      "tag": "h6",
                    },
                    "ref": null,
                    "rendered": Array [
                      1,
                      " x ",
                      2,
                    ],
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "card-text",
                      "dangerouslySetInnerHTML": Object {
                        "__html": "",
                      },
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "div",
                  },
                ],
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": "button",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
